### ISSUE TITLE
CI/CD: hitting DockerHub pull rate limit a.k.a. TOOMANYREQUESTS

### ISSUE THREAD

**Issue by uncleDecart** [lf-edge/eve#4008]
This is known CI/CD issue affecting our building and testing pipeline both on Eden and EVE side. DockerHub has pull rate limits as of November 2020, BuildJet runners, which are currently used in Eden and EVE. In their documentation, BuildJet recommends [here](https://github.com/lf-edge/eve/blob/master/pkg/pillar/cmd/msrv/msrv_test.go#L140) to authenticate with Docker account to mitigate pull rate limits, since anonymous DockerHub pulls are shared across all the actions run on BuildJet. That would give us 200 pulls daily, compared to 100 daily pulls per anonymous address shared with others. We should authenticate, it'll help us, but this solution is still not sustainable. Note that this pull rate does not apply to some containers, such as LF-Edge containers. Currently known affected images are Redis, registry, alpine, nginx. Packing and distributing them under lf-edge would solve the problem, but I'm not sure we can do that because of licensing.

Previously migrating to ghcr was a viable option, since it didn't have pull rate limit   According to this [discussion](https://github.com/orgs/community/discussions/49671) ghcr now also has pull rate limit.

I'm opening this issue to keep track on this problem.

**Comment by giggsoff** [comment]
As I remember previously we solved the problem having [sposored oss bage](https://docs.docker.com/docker-hub/download-rate-limit/#definition-of-limits). Is it still help us to not hit the limit at least for the part of repos?

**Comment by uncleDecart** [comment]
> As I remember previously we solved the problem having [sposored oss bage](https://docs.docker.com/docker-hub/download-rate-limit/#definition-of-limits). Is it still help us to not hit the limit at least for the part of repos?

@giggsoff  redis, registry, alpine, nginx can't have sponsored oss badge. In theory, we can repackage them under lf-edge, but I think we might have a problem with license

**Comment by OhmSpectator** [comment]
Can we pull the images from some local registry, at least for tests?.. =(

**Comment by deitch** [comment]
Distributing them under OSS will create all sorts of licensing issues, as well as potential issues for lf-edge itself.

What about storing them somehow on our own? Could we package them up as tar files and then load them into local docker for tests? So we pull them exactly once, save them as tar, and then use them in the tests? I am not quite sure where we would store them. GHA artifacts don't quite work, as they have limits. Maybe ghcr like you suggested, or Docker Hub, but limited to pulling just for these actions? The problem is that these are all OSS, so they would need to be open to anyone, wouldn't they, recreating the licensing and distribution issues?

**Comment by uncleDecart** [comment]
> What about storing them somehow on our own? 

Where would you store them? EVE is an open source product, so storage should not be affiliated with any commercial company AFAIU 

> Maybe ghcr like you suggested, or Docker Hub, but limited to pulling just for these actions

Yes, ghcr now have pull rate limit. I think GitHub artefacts don't have pull rate limit, so packaging and sharing them would help us. But if I get it right, docker pricing policy is 5 dollars per month for 5000 pulls per day 

> The problem is that these are all OSS, so they would need to be open to anyone, wouldn't they, recreating the licensing and distribution issues?

Exactly

**Comment by rene** [comment]
@deitch, your fixes related to this issue are merged, are we good to close it?

**Comment by milan-zededa** [comment]
> @deitch, your fixes related to this issue are merged, are we good to close it?

Note that we still do not use the paid docker account for eden tests and may therefore hit this problem. However, I'm not sure if it is possible to access secrets with `pull_request_review`...

**Comment by uncleDecart** [comment]
> However, I'm not sure if it is possible to work around the issue of not having access to secrets with pull_request_review...

@milan-zededa we can pass secrets as input parameters :) I don't know how insecure this approach is, but it should work AFAIU

**Comment by deitch** [comment]
> are we good to close it?

Let's leave it until we are sure that the eden question is resolved?

> paid docker account for eden tests

By which you mean tests on PRs against eden? Or the eden tests on PRs in this repo? I think you mean the latter, but checking to be sure.

The eden tests should not be pulling much in terms of images, are they? I think @yash-zededa was checking into what docker hub images we depend upon there.

**Comment by yash-zededa** [comment]
> > However, I'm not sure if it is possible to work around the issue of not having access to secrets with pull_request_review...
> 
> @milan-zededa we can pass secrets as input parameters :) I don't know how insecure this approach is, but it should work AFAIU

If we change the `pull_request` event to `pull_request_target` and set `secrets inherit` to true in the `eden.yml` file under `lf-edge`, the Eden tests should receive the necessary credentials (non-null) to perform a Docker login. I will create a PR for this

**Comment by deitch** [comment]
Are there security concerns @yash-zededa ?

**Comment by yash-zededa** [comment]
@deitch @milan-zededa  @uncleDecart 

I am considering adding a simple run section in the `eden.yml` to check the status of the Docker pull rate limits. Do you have any suggestions? Is this a good idea, a bad idea, or unnecessary?

**Comment by milan-zededa** [comment]
> > > However, I'm not sure if it is possible to work around the issue of not having access to secrets with pull_request_review...
> > 
> > 
> > @milan-zededa we can pass secrets as input parameters :) I don't know how insecure this approach is, but it should work AFAIU
> 
> If we change the `pull_request` event to `pull_request_target` and set `secrets inherit` to true in the `eden.yml` file under `lf-edge`, the Eden tests should receive the necessary credentials (non-null) to perform a Docker login. I will create a PR for this

Will we get the same behavior as with pull_request_**review** (triggered on approve etc.)

**Comment by yash-zededa** [comment]
> > > > However, I'm not sure if it is possible to work around the issue of not having access to secrets with pull_request_review...
> > > 
> > > 
> > > @milan-zededa we can pass secrets as input parameters :) I don't know how insecure this approach is, but it should work AFAIU
> > 
> > 
> > If we change the `pull_request` event to `pull_request_target` and set `secrets inherit` to true in the `eden.yml` file under `lf-edge`, the Eden tests should receive the necessary credentials (non-null) to perform a Docker login. I will create a PR for this
> 
> Will we get the same behavior as with pull_request_**review** (triggered on approve etc.)?

I think so. There shouldn't be any issue with triggering on approval, but we need to check the context that will be passed to Eden. Since with `pull_request_target` the default context would change, we should be able to manage that.

**Comment by uncleDecart** [comment]
EVE has sponsored OSS badge on docker hub, therefore not affected @giggsoff

**Comment by giggsoff** [comment]
> EVE has sponsored OSS badge on docker hub, therefore not affected @giggsoff

I know, but my comment is not about EVE-OS image, but about containers we run inside EVE-OS during tests (see example in my previous comment).

**Comment by uncleDecart** [comment]
Ahhh, apologies, you're right... We can technically copy docker registry throughout tests also, or implement credential ingestion via env variables

**Comment by deitch** [comment]
Those definitely will be subject to rate limits. Even with our new 5,000-pull-per-day account, we eventually will hit it. We run multiple eden tests on each PR (matrix of variants), multiple PRs per day, etc.

What can we do to avoid the issue specifically there? We have much more control over things within eden tests. Could we start by taking an inventory of every image that eden tests pull?

Or does this deserve its own issue?

**Comment by giggsoff** [comment]
> What can we do to avoid the issue specifically there? 

Few thoughts:
1. Pass credentials into EVE-OS config, generated by Eden (it is not supported by Eden's CLI now).
2. Pull images into local registry and points EVE-OS onto it in tests we want to unblock (additional step to pull image into local registry and additional flag in pod deploy. Both can be applied in test,  see the sample [here).](https://github.com/lf-edge/eden/blob/ffbccba605c3f3782084b43a9ee4664e1a18164f/tests/registry/testdata/registry_test.txt#L11-L17, but not sure about auth during `eden registry load` command.)

> Or does this deserve its own issue?

Assume it should be in Eden repo at least.

**Comment by deitch** [comment]
> Pass credentials into EVE-OS config, generated by Eden (it is not supported by Eden's CLI now).

I actually don't know what happens when you do a pull from within eden, when running inside a GHA run, where it stored the credentials already via an earlier `docker login` command. Does eden's docker interactions consume the default locations?

Either way, it is good to have eden be able to be configured for OCI registry credentials, both using the default location and explicitly configured.

> Pull images into local registry and points EVE-OS onto it in tests we want to unblock

That doesn't change anything, other than where credentials get consumed. Either way, you would have one pull per image (nginx or alpine or redis, etc.) per pipeline run. Because each run is its own VM, it has no knowledge of the others.

> Assume it should be in Eden repo at least

Yeah, you are right. This deserves its own issue in eden. I will start one there and we can close this one.

**Comment by deitch** [comment]
I linked to the 2 new issues. Feel free to edit/add/etc.

**Comment by deitch** [comment]
> we will use logged in limit of 200 pulls

We already have 5,000 pulls/day limit.

**Comment by uncleDecart** [comment]
Okay, 5000 pulls/day limit. Point is it's limited :) For now I guess it's not that urgent, but knowing where to fix once it's stopped working is a great thing :D

**Comment by deitch** [comment]
> we are gonna reuse images between tests

There are 2 things mixed together here, each of which is valuable on its own:

* reusing images between tests in a single PR, e.g. matrix that runs similar tests on different platforms or different configurations
* reusing images between test runs, e.g. a test on a single PR fails, someone fixes it, and we rerun it, then do it 5 times. Except for the thing being changed itself - eve - everything else should be identical; or 2 distinct PRs, each of which has to pull the same images.

**Comment by deitch** [comment]
> where to fix once it's stopped working is a great thing 

👍

**Comment by deitch** [comment]
> not that urgent

This is the only thing with which I disagree. We put in a _lot_ of effort in the last week or two to clean a lot up. Let's keep this going for a few more days to finish it. 😄

**Comment by uncleDecart** [comment]
Agreed, it's better done properly. But once this 2 issues are solved on Eden side we're good (for now), right? Would be actually nice to count those pull rates throughout GitHub action execution, or at least eden test... That way we can estimate when we will hit pull rate limit 🤔

**Comment by yash-zededa** [comment]
> > I am considering adding a simple run section in the eden.yml to check the status of the Docker pull rate limits. Do you have any suggestions? Is this a good idea, a bad idea, or unnecessary?
> 
> Definitely not a bad idea! It might be unnecessary, as we can hit it anywhere, not just in eden.
> 
> If you want to check it at the beginning of the entire eve PR flow, some kind of `report` at the beginning of `build.yml` or even its own workflow file, not matrixed, before everything else, that would be helpful.
> 
> They describe how to [here](https://docs.docker.com/docker-hub/download-rate-limit/#how-can-i-check-my-current-rate)

I will add a run section that will check the rate limits just for our information at the beginning of the PR

**Comment by uncleDecart** [comment]
Can we check pull rate limits before and after running tests @yash-zededa? That way we can calculate how many pulls we do during testing, building, etc.

**Comment by rene** [comment]
Closing since this is already fixed.
